interface.go

```diff
@@ -0,0 +1,74 @@
+type Interface interface {
+
+	// PublishLocal appends adv to the locally stored advertisement chain and returns the corresponding CID to it.
+	// This function does not publish the changes to the advertisement chain onto gossip pubsub channel.
+	// Use Publish instead if indexer nodes must be made aware of the appended advertisement.
+	//
+	// See: Publish.
+	PublishLocal(ctx context.Context, adv schema.Advertisement) (cid.Cid, error)
+
+	// Publish appends adv to the locally stored advertisement chain, and publishes the new advertisement onto gossip pubsub.
+	// The CID returned represents the ID of the advertisement appended to the chain.
+	Publish(ctx context.Context, adv schema.Advertisement) (cid.Cid, error)
+
+	// RegisterCallback registers the callback used by the provider to look up a list of multihashes by LookupKey.
+	// Only a single callback is supported; repeated calls to this function will replace the previous callback.
+	RegisterCallback(cb Callback)
+
+	// NotifyPut sginals to the provider that the list of multihashes looked up by the given key are available.
+	// The given key is then used to look up the list of multihashes via Callback.
+	// An advertisement is then generated, appended to the chain of advertisements and published onto the gossip pubsub channel.
+	// Therefore, a Callback must be registered prior to using this function.
+	//
+	// The metadata is serialized data that provides hits about how to retrieve data and is entirely protocol dependant.
+	// The metadata may be nil and is optional.
+	// If specified, the metadata value must start with a uvarint that correspons to the indexer.Value protocol ID.
+	//
+	// This function returns the ID of the advertisement published.
+	//
+	// See: https://pkg.go.dev/github.com/filecoin-project/go-indexer-core
+	// TODO: update pkg link above once encodeMetadata is exported and docs are updated.
+	NotifyPut(ctx context.Context, key LookupKey, metadata []byte) (cid.Cid, error)
+
+	// NotifyRemove sginals to the provider that the multihashes that corresponded to the given key are no longer available.
+	// An advertisement is then generated, appended to the chain of advertisements and published onto the gossip pubsub channel.
+	// The given key must have previously been put via NotifyPut.
+	//
+	// This function returns the ID of the advertisement published.
+	NotifyRemove(ctx context.Context, key LookupKey) (cid.Cid, error)
+
+	// GetAdv gets the advertisement that corresponds to the given id.
+	GetAdv(ctx context.Context, id cid.Cid) (schema.Advertisement, error)
+
+	// GetLatestAdv gets the latest advertisement on this provider's advertisement chain and the CID to which it corresponds.
+	GetLatestAdv(ctx context.Context) (cid.Cid, schema.Advertisement, error)
+
+	// Close closes this provider, discarding all resources occupied by it.
+	// After calling this function the provider is no longer available for use.
+	Close(ctx context.Context) error
```

Comment:
Since we're here, I think we should rename this to Shutdown, given that it takes a context and presumably blocks until all goroutines have stopped. A Close operation tends to be immediate and not take a context.